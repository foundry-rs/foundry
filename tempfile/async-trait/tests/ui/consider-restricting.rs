// https://github.com/rust-lang/rust/issues/93828

use async_trait::async_trait;

pub trait IntoUrl {}

#[async_trait]
pub trait ClientExt {
    async fn publish<T: IntoUrl>(&self, url: T);
}

struct Client;

#[async_trait]
impl ClientExt for Client {
    async fn publish<T: IntoUrl>(&self, url: T) {}
}

struct Client2;

#[async_trait]
impl ClientExt for Client2 {
    async fn publish<T>(&self, url: T) {}
}

fn main() {}
