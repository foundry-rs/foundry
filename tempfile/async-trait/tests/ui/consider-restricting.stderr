error: future cannot be sent between threads safely
  --> tests/ui/consider-restricting.rs:16:49
   |
16 |     async fn publish<T: IntoUrl>(&self, url: T) {}
   |                                                 ^^ future created by async block is not `Send`
   |
note: captured value is not `Send`
  --> tests/ui/consider-restricting.rs:16:41
   |
16 |     async fn publish<T: IntoUrl>(&self, url: T) {}
   |                                         ^^^ has type `T` which is not `Send`
   = note: required for the cast from `Pin<Box<{async block@$DIR/tests/ui/consider-restricting.rs:16:49: 16:51}>>` to `Pin<Box<(dyn Future<Output = ()> + Send + 'async_trait)>>`
help: consider further restricting type parameter `T` with trait `Send`
   |
16 |     async fn publish<T: IntoUrl + std::marker::Send>(&self, url: T) {}
   |                                 +++++++++++++++++++

error: future cannot be sent between threads safely
  --> tests/ui/consider-restricting.rs:23:40
   |
23 |     async fn publish<T>(&self, url: T) {}
   |                                        ^^ future created by async block is not `Send`
   |
note: captured value is not `Send`
  --> tests/ui/consider-restricting.rs:23:32
   |
23 |     async fn publish<T>(&self, url: T) {}
   |                                ^^^ has type `T` which is not `Send`
   = note: required for the cast from `Pin<Box<{async block@$DIR/tests/ui/consider-restricting.rs:23:40: 23:42}>>` to `Pin<Box<(dyn Future<Output = ()> + Send + 'async_trait)>>`
help: consider further restricting type parameter `T` with trait `Send`
   |
23 |     async fn publish<T + std::marker::Send>(&self, url: T) {}
   |                        +++++++++++++++++++
